### PR TITLE
homework-5 (with bonus)

### PR DESCRIPTION
**Pull request by Efler**
#### Github:
* обновление метаданных/настроек репозитория
* новый пуш в репозитории
* новая ветка в репозитории
* удаление ветки в репозитории
#### Stackoverflow:
* Новый ответ на вопрос
* Новый комментарий к вопросу/ответу
* Ответ был принят автором вопроса
* Состояние вопроса было изменено

### CODE REVIEW

**Comment by Efler** [comment]
### Code Coverage
|Overall Project|41.78% **`-52.51%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|42.08%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[LinkUpdateScheduler.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fscheduler%2FLinkUpdateScheduler.java)|100%|:green_apple:|
|[ChatDao.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fdomain%2Fjdbc%2Fdao%2FChatDao.java)|100%|:green_apple:|
|[TrackingDao.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fdomain%2Fjdbc%2Fdao%2FTrackingDao.java)|100%|:green_apple:|
|[JpaLinkService.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fservice%2Fjpa%2FJpaLinkService.java)|100%|:green_apple:|
|[StackoverflowClient.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fclient%2FStackoverflowClient.java)|94.7% **`-5.3%`**|:green_apple:|
|[LinkDao.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fdomain%2Fjdbc%2Fdao%2FLinkDao.java)|91.54% **`-8.46%`**|:green_apple:|
|[JpaTgChatService.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fservice%2Fjpa%2FJpaTgChatService.java)|83.48% **`-16.52%`**|:green_apple:|
|[BranchDao.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fdomain%2Fjdbc%2Fdao%2FBranchDao.java)|83.26% **`-16.74%`**|:green_apple:|
|[GithubClient.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fclient%2FGithubClient.java)|75.12% **`-24.88%`**|:green_apple:|
|[JpaLinkUpdateService.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fservice%2Fjpa%2FJpaLinkUpdateService.java)|58.76% **`-41.24%`**|:green_apple:|
|[ScrapperApplication.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2FScrapperApplication.java)|37.5%|:green_apple:|
|[TgChatController.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fcontroller%2FTgChatController.java)|27.27% **`-72.73%`**|:x:|
|[BotClient.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fclient%2FBotClient.java)|19.09% **`-54.55%`**|:x:|
|[SecurityService.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fservice%2Fsecurity%2FSecurityService.java)|16.67% **`-83.33%`**|:x:|
|[LinksController.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fcontroller%2FLinksController.java)|5.83% **`-80.58%`**|:x:|
|[RestExceptionHandler.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fadvice%2FRestExceptionHandler.java)|1.35%|:green_apple:|
|[ScrapperQueueProducer.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fservice%2Fkafka%2FScrapperQueueProducer.java)|0%|:x:|
|[JdbcLinkService.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fservice%2Fjdbc%2FJdbcLinkService.java)|0%|:x:|
|[JdbcLinkUpdateService.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fservice%2Fjdbc%2FJdbcLinkUpdateService.java)|0%|:x:|
|[JdbcTgChatService.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fservice%2Fjdbc%2FJdbcTgChatService.java)|0%|:x:|
|[JooqTgChatService.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fservice%2Fjooq%2FJooqTgChatService.java)|0%|:x:|
|[JooqLinkUpdateService.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fservice%2Fjooq%2FJooqLinkUpdateService.java)|0%|:x:|
|[JooqLinkService.java](https://github.com/Efler/telegram-scrapper-bot/blob/7970dbba00150c9a0747108cba0657498d2094c3/scrapper%2Fsrc%2Fmain%2Fjava%2Fedu%2Feflerrr%2Fscrapper%2Fservice%2Fjooq%2FJooqLinkService.java)|0%|:x:|

File: scrapper/src/main/java/edu/eflerrr/scrapper/configuration/TimeConstants.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+import java.time.OffsetDateTime;
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public final class TimeConstants {
+
+    public final static OffsetDateTime MIN_DATE_TIME =
+        OffsetDateTime.parse("0001-01-01T00:00:00Z");
```
Comment (Betorov) [inline scrapper/src/main/java/edu/eflerrr/scrapper/configuration/TimeConstants.java]: OffsetDateTime.now()?

File: scrapper/src/main/java/edu/eflerrr/scrapper/domain/jdbc/dao/BranchDao.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        return jdbcTemplate.query(sql, (rs, rowNum) ->
+            new Branch(
+                rs.getLong("id"),
+                rs.getString("repository_owner"),
+                rs.getString("repository_name"),
+                rs.getString("branch_name"),
+                rs.getObject("last_commit_time", OffsetDateTime.class)
+            ), owner, name
+        );
+    }
```
Comment (Betorov) [inline scrapper/src/main/java/edu/eflerrr/scrapper/domain/jdbc/dao/BranchDao.java]: BranchDao?

В задании 5-ом хотели реализовать через интерфейс, BranchDao, JdbcBranchDao, JooqBranchDao и т.д..

File: scrapper/src/main/java/edu/eflerrr/scrapper/domain/jooq/DefaultCatalog.java
Comment (Betorov) [inline scrapper/src/main/java/edu/eflerrr/scrapper/domain/jooq/DefaultCatalog.java:1]: jooq, Предлагаю создать отдельный модуль где будет лежать вся модель. ? Почему, потому что ее часто приходится перегенирировать. И будет аккуратнее выглядеть

**Review by Betorov — APPROVED** [review]

File: scrapper/src/main/java/edu/eflerrr/scrapper/domain/jdbc/dao/BranchDao.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            """;
+        var rowCount = jdbcTemplate.queryForObject(
+            countSql,
+            Integer.class,
+            branch.getRepositoryOwner(),
+            branch.getRepositoryName(),
+            branch.getBranchName()
+        );
+        return rowCount != null && rowCount == 1;
+    }
```
Comment (Betorov) [inline scrapper/src/main/java/edu/eflerrr/scrapper/domain/jdbc/dao/BranchDao.java]: У всех какая-то общая проблема. Создают транзакцию избыточно. 
На селекте то она зачем?

File: scrapper/src/main/java/edu/eflerrr/scrapper/domain/jdbc/dao/BranchDao.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        jdbcTemplate.update(
+            sql,
+            branch.getRepositoryOwner(),
+            branch.getRepositoryName(),
+            branch.getBranchName(),
+            branch.getLastCommitTime()
+        );
+    }
+
+    @Transactional
```
Comment (Betorov) [inline scrapper/src/main/java/edu/eflerrr/scrapper/domain/jdbc/dao/BranchDao.java]: В одной бизнес-логике используется несколько транзакций, точно он так нужно? Может мы должны удалить в одном месте и в другом сразу?
Транзакция выглядит избыточной.

Обычно ее вставляют когда в бизнес-логике идет работа с БД (а тут с несколькими таблицами сразу)

**Comment by Efler** [comment]
Здравствуйте! Если что, отписал в телеграм по поводу правок

File: scrapper/src/main/java/edu/eflerrr/scrapper/configuration/TimeConstants.java (hunk 1 above)
Comment (Efler) [reply scrapper/src/main/java/edu/eflerrr/scrapper/configuration/TimeConstants.java]: Подправил на OffsetDateTime.now()

File: scrapper/src/main/java/edu/eflerrr/scrapper/domain/jdbc/dao/BranchDao.java (hunk 2 above)
Comment (Efler) [reply scrapper/src/main/java/edu/eflerrr/scrapper/domain/jdbc/dao/BranchDao.java]: Обговорили в телеграме

File: scrapper/src/main/java/edu/eflerrr/scrapper/domain/jooq/DefaultCatalog.java
Comment (Efler) [reply scrapper/src/main/java/edu/eflerrr/scrapper/domain/jooq/DefaultCatalog.java:1]: Сделано!

File: scrapper/src/main/java/edu/eflerrr/scrapper/domain/jdbc/dao/BranchDao.java (hunk 3 above)
Comment (Efler) [reply scrapper/src/main/java/edu/eflerrr/scrapper/domain/jdbc/dao/BranchDao.java]: Подправил

File: scrapper/src/main/java/edu/eflerrr/scrapper/domain/jdbc/dao/BranchDao.java (hunk 4 above)
Comment (Efler) [reply scrapper/src/main/java/edu/eflerrr/scrapper/domain/jdbc/dao/BranchDao.java]: Исправлено!

**Comment by Efler** [comment]
### Схема scrapper

scrapper hierarchy

**Pull request merged**